src/server.h

```diff
@@ -1757,10 +1757,10 @@ struct redisServer {
     rax *errors;                /* Errors table */
     int errors_enabled;         /* If true, errorstats is enabled, and we will add new errors. */
     unsigned int lruclock; /* Clock for LRU eviction */
-    volatile sig_atomic_t shutdown_asap; /* Shutdown ordered by signal handler. */
-    volatile sig_atomic_t crashing;      /* Server is crashing report. */
+    redisAtomic int shutdown_asap; /* Shutdown ordered by signal handler. */
+    redisAtomic int crashing;      /* Server is crashing report. */
```

Comment:
I wonder if we are getting this right. Documentation mentions `volatile sig_atomic_t` is guaranteed to work in a signal handler. I understand the reason is need for compiler and CPU barriers. So, with redisAtomic, we will also get the same properties. Though, now, the access to these variables should use atomicSet()/atomicGet() functions.

Reply:
i think sanitizer doesn't report errors after changing to `redisAtomic`,  the reason is that the implementation of `redisAtomic` is `_Atomic` for new compiler version(newer than 4.9), so `++` or `--` of them are `memory_order_seq_cst` semantics even we don't use `atomicSet()/atomicGet()`

Reply:
I was also under the impression that `volatile sig_atomic_t` is enough nevertheless thread sanitizer still flags it as problematic.
@ShooterIT is right, `_Atomic` shouldn't need `atomicSet()/atomicGet()`
I'm open to suggestions here; redisAtomic works and I didn't come up with any better alternatives.

Reply:
My question is what happens on an older compiler without `_Atomic`? Because `_Atomic`  was introduced in C11. Now, read/write to these variables are not guaranteed to be atomic and there is no visibility guarantee without atomicGet()/atomicSet(), no?

Reply:
@tezc you are right, i just said why this code can work, as you said, the sanitizer may report errors if we use gcc 4.8. we should use `atomicSet()/atomicGet()` if we want to this change compatible

> I was also under the impression that volatile sig_atomic_t is enough nevertheless thread sanitizer still flags it as problematic.

I agree volatile sig_atomic_t is enough is enough, maybe  it is a false positive, sanitizer can't distinguish the signal hander and thread

Reply:
I'm concerned about regular build rather than TSAN false reports. Code may not work correctly without atomicGet() / atomicSet() on an older compiler.

Reply:
I think the story behind "`volatile sig_atomic_t` is enough" is that in single threaded setup, the signal handler is called while the main thread is suspended. But with many threads the kernel can decide calling the the handler on any other thread (modulo `pthread_sigmask()`) while the main thread is still running.
@ShooterIT suggested using `pthread_sigmask()` to ensure that the signal handler is always called on the main thread. Unfortunately I couldn't make this approach work, because I have no control over the sigmask of threads created by modules.
Instead, added `atomicGet/Set()` to the appropriate places so that we support compilers without C11's _Atomic, as @tezc suggested.
Can you please take another look?

Reply:
yeah, looks like the issue has been there forever but probably very unlikely to happen. Okay, so we have atomicGet()/atomicSet() now and we should be safe.